examples/integrations/hf_trainer_api/det_callback.py

```diff
@@ -0,0 +1,243 @@
+        self.searcher_unit = list(searcher_config["max_length"].keys())[0]
+        self.searcher_max_length = list(searcher_config["max_length"].values())[0]
+        self.searcher_ops = self.core_context.searcher.operations()
+        self.current_op = next(self.searcher_ops)
+
+        self._check_searcher_compatibility(args)
+
+    def on_log(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        logs=None,
+        **kwargs,
+    ):
+        if state.is_world_process_zero:
+            metrics, metric_type = self._get_metrics(logs)
+            if metric_type == TRAIN:
+                # Prevents reporting metrics for the same step twice. This happens after
+                # training is completed and average training metrics are reported with
+                # the same step as the in-progress training metrics.
+                if self.last_metrics["train_step"] != state.global_step:
+                    self.core_context.train.report_training_metrics(
+                        steps_completed=state.global_step, metrics=metrics
+                    )
+                    metrics["train_step"] = state.global_step
+
+            elif metric_type == EVAL:
+                # Prevents reporting metrics for the same step twice. This happens when
+                # after-training evaluation is completed, and it is reported with the same
+                # step as the last during-training evaluation.
+                if self.last_metrics["eval_step"] != state.global_step:
+                    self.core_context.train.report_validation_metrics(
+                        steps_completed=state.global_step, metrics=metrics
+                    )
+                    metrics["eval_step"] = state.global_step
+                    self.last_metrics.update(metrics)
+            else:
+                logging.warning(f"Metrics not reported: metric type = {metric_type}.")
+
+            self.last_metrics.update(metrics)
+
+        if self.core_context.preempt.should_preempt():
+            control.should_save = True
+
+    def _get_metrics(self, logs: typing.Dict) -> typing.Tuple[typing.Dict, str]:
+        metrics = logs
+        metric_type = get_metric_type(logs)
+        if self.filter_metrics:
+            metrics = {}
+            for k, v in logs.items():
+                if any(m in k for m in self.filter_metrics) is True:
+                    metrics[k] = v
+
+        return metrics, metric_type
+
+    def on_save(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        **kwargs,
+    ):
+        info = det.get_cluster_info()
+        assert info
+
+        # local_path is where HF Trainer saves model and tokenizer.
+        local_path = os.path.join(args.output_dir, f"checkpoint-{state.global_step}")
+
+        if state.is_world_process_zero:
+            if self.user_checkpoint_metadata is not None:
```

Comment:
can you provide an example usage of `user_checkpoint_metadata`?  with an overwritten `_on_save_user_data`?